test/cpp/end2end/mtls_end2end_test.cc

```diff
@@ -0,0 +1,409 @@
+        server_pair.private_key = server_key;
+        server_pair.certificate_chain = server_certs;
+        std::vector<experimental::IdentityKeyCertPair> server_pair_list;
+        server_pair_list.emplace_back(server_pair);
+        server_certificate_provider =
+            std::make_shared<experimental::StaticDataCertificateProvider>(
+                root_certs, server_pair_list);
+        break;
+      }
+      case SecurityPrimitives::FILE_PROVIDER: {
+        server_certificate_provider =
+            std::make_shared<FileWatcherCertificateProvider>(
+                SERVER_KEY_PATH, SERVER_CERT_PATH, CA_CERT_PATH, 1);
+        break;
+      }
+    }
+    // Set up server certificate verifier.
+    std::shared_ptr<experimental::CertificateVerifier>
+        server_certificate_verifier;
+    switch (GetParam().server_verifier_type()) {
+      case SecurityPrimitives::EXTERNAL_SYNC_VERIFIER: {
+        server_certificate_verifier =
+            experimental::ExternalCertificateVerifier::Create<
+                SyncCertificateVerifier>(true);
+        break;
+      }
+      case SecurityPrimitives::EXTERNAL_ASYNC_VERIFIER: {
+        server_certificate_verifier =
+            experimental::ExternalCertificateVerifier::Create<
+                AsyncCertificateVerifier>(true);
+        break;
+      }
+      case SecurityPrimitives::HOSTNAME_VERIFIER: {
+        server_certificate_verifier =
+            std::make_shared<experimental::HostNameCertificateVerifier>();
+        break;
+      }
+      default: {
+        break;
+      }
+    }
+    // Build the server and add listening ports.
+    if (GetParam().num_listening_ports() > 0) {
+      ports_.resize(GetParam().num_listening_ports(), 0);
+    }
+    for (int i = 0; i < GetParam().num_listening_ports(); i++) {
+      // Configure tls credential options for each port
+      TlsServerCredentialsOptions server_creds_options(
+          server_certificate_provider);
+      server_creds_options.set_cert_request_type(
+          GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY);
+      server_creds_options.watch_identity_key_cert_pairs();
+      server_creds_options.watch_root_certs();
+      if (GetParam().server_verifier_type() !=
+          SecurityPrimitives::DEFAULT_VERIFIER) {
+        server_creds_options.set_certificate_verifier(
+            server_certificate_verifier);
+      }
+      builder.AddListeningPort(
+          "0.0.0.0:0",
+          ::grpc::experimental::TlsServerCredentials(server_creds_options),
+          &ports_[i]);
+    }
+    builder.RegisterService(&service_);
+    server_ = builder.BuildAndStart();
+    ASSERT_NE(nullptr, server_);
+    server_thread_ = std::thread(&AdvancedTlsEnd2EndTest::RunServerLoop, this);
+    for (int i = 0; i < GetParam().num_listening_ports(); i++) {
+      ASSERT_NE(0, ports_[i]);
+      server_addresses_.push_back(absl::StrCat("localhost:", ports_[i]));
+      // Set up client certificate provider.
```

Comment:
Same thing here: All of the client creds setup can be done just once, not once per port.  We can use the same `TlsCredentials` object for multiple channels.